File: VideoStreamingBackend/src/application/use-cases/NegotiateViewer.ts
/**
 * Use case: Negotiate a WebRTC viewer connection.
 *
 * This use case is called from the signaling WebSocket handler.
 * It delegates transport/consumer creation to the StreamManager → SFU.
 */
import { StreamManager } from '../../infrastructure/streaming/StreamManager';
import { ICameraService } from '../../domain/services';
import { StreamNotFoundError, ProducerNotFoundError } from '../../domain/errors';
import { logger } from '../../infrastructure/logging/logger';
import type {
  RtpCapabilities,
  WebRtcTransportParams,
  DtlsParameters,
  ConsumerParams,
} from '../../domain/services';

export interface NegotiateViewerInput {
  cameraId: string;
  userId: string;
  token: string;
}

export class NegotiateViewerUseCase {
  constructor(
    private readonly streamManager: StreamManager,
    private readonly cameraService: ICameraService,
  ) {}

  /**
   * Ensure the camera stream is active and return Router RTP capabilities.
   * Called when a viewer first connects.
   */
  async ensureStreamAndGetCapabilities(input: NegotiateViewerInput): Promise<{
    rtpCapabilities: RtpCapabilities;
    videoProducerId: string | null;
    audioProducerId: string | null;
  }> {
    const { cameraId, userId, token } = input;

    // Auto-start ingest if not running
    if (!this.streamManager.isStreaming(cameraId)) {
      logger.info(`[NegotiateViewer] Auto-starting stream for ${cameraId}`);
      const camera = await this.cameraService.getCamera(cameraId, token);
      await this.streamManager.startStream(cameraId, camera.ownerUserId, camera.streamUrl, {
        fps: camera.fps,
      });
    }

    const rtpCapabilities = this.streamManager.getRouterRtpCapabilities(cameraId);
    if (!rtpCapabilities) {
      throw new StreamNotFoundError(cameraId);
    }

    const session = this.streamManager.getSession(cameraId);

    return {
      rtpCapabilities,
      videoProducerId: session?.videoProducerId ?? null,
      audioProducerId: session?.audioProducerId ?? null,
    };
  }

  /**
   * Create a WebRtcTransport for the viewer.
   */
  async createTransport(cameraId: string): Promise<WebRtcTransportParams> {
    return this.streamManager.createViewerTransport(cameraId);
  }

  /**
   * Connect the viewer's transport (DTLS handshake).
   */
  async connectTransport(
    cameraId: string,
    transportId: string,
    dtlsParameters: DtlsParameters,
  ): Promise<void> {
    return this.streamManager.connectViewerTransport(cameraId, transportId, dtlsParameters);
  }

  /**
   * Create a Consumer so the viewer receives media.
   */
  async consume(
    cameraId: string,
    transportId: string,
    producerId: string,
    rtpCapabilities: RtpCapabilities,
  ): Promise<ConsumerParams> {
    if (!producerId) {
      throw new ProducerNotFoundError(cameraId);
    }
    return this.streamManager.createConsumer(cameraId, transportId, producerId, rtpCapabilities);
  }

  /**
   * Resume a paused Consumer (starts media flow).
   */
  async resumeConsumer(cameraId: string, consumerId: string): Promise<void> {
    return this.streamManager.resumeConsumer(cameraId, consumerId);
  }

  /**
   * Clean up when viewer disconnects.
   */
  disconnect(cameraId: string, transportId: string): void {
    this.streamManager.closeViewerTransport(cameraId, transportId);
    this.streamManager.removeViewer(cameraId);
  }
}

File: StreamingBackend/internal/application/usecases/negotiate_viewer.go
package usecases

import (
	"github.com/google/uuid"
	"github.com/pion/webrtc/v4"
	"github.com/rs/zerolog/log"
	"github.com/vigileye/streaming-backend/internal/domain/services"
	"github.com/vigileye/streaming-backend/internal/infrastructure/streaming"
)

// NegotiateViewerInput contains the input for WebRTC negotiation.
type NegotiateViewerInput struct {
	CameraID string
	UserID   string
	Token    string
}

// NegotiateViewerUseCase handles WebRTC viewer negotiation.
type NegotiateViewerUseCase struct {
	streamManager *streaming.StreamManager
	cameraService services.CameraService
}

// NewNegotiateViewerUseCase creates a new NegotiateViewerUseCase.
func NewNegotiateViewerUseCase(sm *streaming.StreamManager, cs services.CameraService) *NegotiateViewerUseCase {
	return &NegotiateViewerUseCase{
		streamManager: sm,
		cameraService: cs,
	}
}

// NegotiateOffer handles a WebRTC SDP offer from a viewer.
// It auto-starts the stream if not already running, then returns the SDP answer.
func (uc *NegotiateViewerUseCase) NegotiateOffer(input NegotiateViewerInput, offer webrtc.SessionDescription) (*webrtc.SessionDescription, string, error) {
	// Auto-start stream if not running
	if !uc.streamManager.IsStreaming(input.CameraID) {
		log.Info().Str("camera_id", input.CameraID).Msg("Auto-starting stream for WebRTC viewer")
		camera, err := uc.cameraService.GetCamera(input.CameraID, input.Token)
		if err != nil {
			return nil, "", err
		}
		if _, err := uc.streamManager.StartStream(input.CameraID, camera.OwnerUserID, camera.StreamURL, nil); err != nil {
			return nil, "", err
		}
	}

	viewerID := uuid.New().String()
	uc.streamManager.AddViewer(input.CameraID)

	answer, err := uc.streamManager.NegotiateWebRTC(input.CameraID, viewerID, offer)
	if err != nil {
		uc.streamManager.RemoveViewer(input.CameraID)
		return nil, "", err
	}

	return answer, viewerID, nil
}

// AddICECandidate adds a trickle ICE candidate for a viewer.
func (uc *NegotiateViewerUseCase) AddICECandidate(cameraID, viewerID string, candidate webrtc.ICECandidateInit) error {
	return uc.streamManager.AddICECandidate(cameraID, viewerID, candidate)
}

// Disconnect cleans up a viewer connection.
func (uc *NegotiateViewerUseCase) Disconnect(cameraID, viewerID string) {
	uc.streamManager.DisconnectViewer(cameraID, viewerID)
}
